GH-119668: expose importlib.machinery.NamespacePath (#119669)
python/cpython: Doc/library/importlib.rst

      :exc:`ImportError` is raised.

.. class:: NamespacePath(name, path, path_finder)

   Represents a namespace package's path.

   It uses the module *name* to find its parent module, and from there it looks
   up the parent's :attr:`module.__path__`.  When this changes, the module's own

>    path is recomputed, using *path_finder*. The initial value is set to *path*.

Reviewer: Can we perhaps mention the type of the path finder? (namely using some class Sphinx directive).

Btw, when you say that the initial value is set to *path* it seems that you are talking about the initial value of the path finder.

In addition, maybe mention that invalidating the caches forces the path to be recomputed (it might not necessarily be the case that the module parent's path has changed).

Finally, could you document the public methods of this new class? For instance there is no insert() method nor remove() as for lists. 

I think an example of how to use it could be useful though I don't know whether we want to burden the docs even more.

Author: > Can we perhaps mention the type of the path finder? (namely using some class Sphinx directive).
> 
> Btw, when you say that the initial value is set to _path_ it seems that you are talking about the initial value of the path finder.
> 
> In addition, maybe mention that invalidating the caches forces the path to be recomputed (it might not necessarily be the case that the module parent's path has changed).

I've rewritten the docs to include this.

> Finally, could you document the public methods of this new class? For instance there is no insert() method nor remove() as for lists.

Using `NamespacePath`'s mutability at runtime is something that requires folks to be familiar with the source, so I don't think we really need to document that.

The main improvement I want to get from the PR is the ability to `isinstance(path, NamespacePath)` safely.

> I think an example of how to use it could be useful though I don't know whether we want to burden the docs even more.

Similarly, I don't really think that is necessary given the advanced nature of the topic.

   For top-level modules, the parent module's path is :data:`sys.path`.

> 

Reviewer: Can we use `.. versionadded:: next` + what's new entry? While this only exposes a private class, I think users should be notified more than just with a blurb.

Reviewer 2: I agree.

Reviewer 3: Maybe `versionchanged`?  You can say something like, prior to 3.15, this class was exposed as the module private `_NamespacePath` class.  With 3.15 and beyond, the class has been publicly exposed as `NamespacePath`.

Author: The attribute is set on `importlib._bootstrap_external`, which is a private module, so I lean towards `versionadded`.

      :exc:`ImportError` is raised.

> .. class:: NamespacePath(name, path, path_finder)

Reviewer: The documentation explains what *name* and *path_finder* is, but it does not explain what *path* is.

>    Represents a namespace package's path.  It uses the module name to find its

Reviewer: ```suggestion
   Represents a namespace package's path.  It uses the module *name* to find its
```

   parent module, and from there it looks up the parent's ``__path__``.  When

>    this changes, the module's own path is recomputed, using path_finder. For

Reviewer: ```suggestion
   this changes, the module's own path is recomputed, using *path_finder*. For
```

>    top-level modules, the parent module's path is :py:data:`sys.path`.

Reviewer: ```suggestion
   top-level modules, the parent module's path is :data:`sys.path`.
```

Same observation.

      :exc:`ImportError` is raised.

.. class:: NamespacePath(name, path, path_finder)

   Represents a :term:`namespace package`'s path (:attr:`module.__path__`).

>    When its value is accessed, if necessary, it will be recomputed.

Reviewer 3: ```suggestion
   When its `__path__` value is accessed it will be recomputed if necessary.
```

   This keeps it in-sync with the global state (:attr:`sys.modules`).

   The *name* argument is the name of the namespace module.

   The *path* argument is the initial path value.

   The *path_finder* argument is the callable used to recompute the path value.

>    It has the same signature as :meth:`MetaPathFinder.find_spec`.

Reviewer 3: It's a little hard to tell from context, but which "it" is "It" referring to?  I'm not sure if this is part of the `path_finder` paragraph or a separate paragraph.

Author: It refers to `path_finder`, I'll rewrite it as "The callable has the same signature as (...)"

>    It has the same signature as :meth:`MetaPathFinder.find_spec`.

Author: ```suggestion
   The *path_finder* argument is the callable used to recompute the path value.
   The callable has the same signature as :meth:`MetaPathFinder.find_spec`.
```

>    The callable has the same signature as :meth:`MetaPathFinder.find_spec`.

Reviewer 2: Note the warning about `MetaPathFinder.find_spec` not being found; probably requires the full name.

```suggestion
   The callable has the same signature as :meth:`importlib.abc.MetaPathFinder.find_spec`.
```
